bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoHandler.java

```diff
@@ -195,19 +227,19 @@ public void onValueReceived(@Nullable String variable, @Nullable String value, @
 
         updateStatus(ThingStatus.ONLINE);
 
+        String oldValue = this.stateMap.get(variable);
         if (variable != null && value != null) {
             this.stateMap.put(variable, value);
         }
 
-        if (getThing().getThingTypeUID().getId().equals("insight")) {
-            String insightParams = stateMap.get("InsightParams");
+        if (value != null && value.length() > 1) {
+            String insightParams = stateMap.get(variable);
```

Comment:
This is a breaking change as you are now accepting different variable types but without a corresponding implementation for that:

```
2022-01-23 14:36:56.825 [DEBUG] [ng.wemo.internal.handler.WemoHandler] - Received pair 'EnergyPerUnitCost':'4|2000|4' (service 'basicevent1') for thing 'wemo:insight:insight1'
2022-01-23 14:36:56.828 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam binaryState 'ON' for device 'wemo:insight:insight1' received
2022-01-23 14:36:56.831 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam lastChangedAt '1970-01-01T01:33:20.000+0100' for device 'wemo:insight:insight1' received
2022-01-23 14:36:56.834 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam lastOnFor '4' for device 'wemo:insight:insight1' received
2022-01-23 14:36:56.836 [ERROR] [port.upnp.internal.UpnpIOServiceImpl] - Participant threw an exception onValueReceived
java.lang.ArrayIndexOutOfBoundsException: Index 3 out of bounds for length 3
	at org.openhab.binding.wemo.internal.handler.WemoHandler.onValueReceived(WemoHandler.java:270) ~[?:?]
	at org.openhab.core.io.transport.upnp.internal.UpnpIOServiceImpl$UpnpSubscriptionCallback.eventReceived(UpnpIOServiceImpl.java:161) [bundleFile:?]
	at org.jupnp.controlpoint.SubscriptionCallback$2.eventReceived(SubscriptionCallback.java:222) [bundleFile:?]
	at org.jupnp.model.gena.RemoteGENASubscription.receive(RemoteGENASubscription.java:114) [bundleFile:?]
	at org.jupnp.protocol.sync.ReceivingEvent$2.run(ReceivingEvent.java:130) [bundleFile:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
```

main branch:

```
2022-01-23 14:38:53.312 [DEBUG] [ng.wemo.internal.handler.WemoHandler] - Received pair 'EnergyPerUnitCost':'4|2000|4' (service 'basicevent1') for thing 'wemo:insight:insight1'
2022-01-23 14:38:53.315 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam binaryState 'OFF' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.318 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam lastChangedAt '2022-01-22T21:44:14.000+0100' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.321 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam lastOnFor '0' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.324 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam onToday '0' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.326 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam onTotal '0' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.328 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam timespan '1209600' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.331 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam averagePower '13 W' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.334 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam currentPower '0 W' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.338 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam energyToday '0 Wh' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.341 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam energyTotal '0 Wh' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.343 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam standByLimit '8 W' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.346 [DEBUG] [ng.wemo.internal.handler.WemoHandler] - Received pair 'BinaryState':'0|1642884254|0|0|0|1209600|13|0|0|0' (service 'basicevent1') for thing 'wemo:insight:insight1'
2022-01-23 14:38:53.348 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam binaryState 'OFF' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.352 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam lastChangedAt '2022-01-22T21:44:14.000+0100' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.354 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam lastOnFor '0' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.357 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam onToday '0' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.359 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam onTotal '0' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.362 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam timespan '1209600' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.364 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam averagePower '13 W' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.367 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam currentPower '0 W' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.371 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam energyToday '0 Wh' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.374 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam energyTotal '0 Wh' for device 'wemo:insight:insight1' received
2022-01-23 14:38:53.376 [TRACE] [ng.wemo.internal.handler.WemoHandler] - New InsightParam standByLimit '8 W' for device 'wemo:insight:insight1' received
```

Please note that everything logged are the previous values on InsightParams from the `stateMap` since this version didn't support **EnergyPerUnitCost** or **BinaryState**. So previously these variables were simply skipped as not supported, and now they are supported but not handled correctly.

Reply:
Did not think of EnergyPerUnitCost. Fix provided...
Interesting, my devices do not report this.....

Reply:
I'm sorry, but this last fix seems very error-prone. Consider these variations of received data:

```
Received pair 'EnergyPerUnitCost':'4|2000|4' (service 'basicevent1') for thing 'wemo:insight:insight1'
Received pair 'BinaryState':'0|1642884254|0|0|0|1209600|13|0|0|0' (service 'basicevent1') for thing 'wemo:insight:insight1'
Received pair 'InsightParams':'1|1632573998|10371148|52746|1174820|1209600|39|26440|28068134|548930560|8000' (service 'insight1') for thing 'wemo:insight:insight2'
```

You are now putting anything into stateMap (previously only InsightParams), so it can be either **EnergyPerUnitCost**, **BinaryState**, **InsightsParams** or something else we don't know yet. You cannot rely on a specific number of parameters or even how they are positioned. You would need to directly support each type, and skip those types not supported. Another view with padding for comparison:

```
EnergyPerUnitCost:
'4|    2000|4'
BinaryState:
'0|1642884254|       0|    0|      0|1209600|13|    0|       0|        0'
InsightParams:
'1|1632573998|10371148|52746|1174820|1209600|39|26440|28068134|548930560|8000'
```

Reply:
No, it is not error prone.
BinaryState can be of length 1 or > 1 for Insight devices. EnergyCostPerUnit will be of length >1, actually size 3.
Following your suggestion, we would need to run the parser if varaiable = InsightParams or BinaryState, but if variable = BinaryState only if length > 1.
This would result in a very complex nesting and hard readability.

Reply:
After a bit more thinkin, we could change the logic to

if ((„InsightParams“.equals(variable) || „BinaryState“.equals(variable)) && value.legth() > 1) {
….
 } else if ( „BinaryState“.equals(variable) && value.length() == 1) {
 ….
}

Reply:
Changed as described....

Reply:
It works now, thanks. Let's go with the bare minimum here. I hope you will not be offended if I take another look at this after merging your PR? :-) I really see a lot of opportunities for simplifications and improved robustness in various places, including this `onValueReceived()` method.

Reply:
> It works now, thanks. Let's go with the bare minimum here. I hope you will not be offended if I take another look at this after merging your PR? :-) I really see a lot of opportunities for simplifications and improved robustness in various places, including this `onValueReceived()` method.

Shure, no problem. If you ways for simplification, go for it.